### PR TITLE
fix: additional profiling data races

### PR DESCRIPTION
**Pull request by indragiek**
## :scroll: Description

While debugging the memory leaks reported in #2980 I found additional data race issues on top of the ones fixed in #3018. This is a larger refactor that fixes those issues but also simplifies the way that synchronization works.

In #3018, there was a change to add locking where `processBacktrace` mutated data structures that could be accessed concurrently. It attempted to separate the code that did not access the shared data structures, and moved all mutating code to a section at the end of the function. However, it missed a few reads from those data structures that also needed to be synchronized (e.g. `frames.count` and `stacks.count`). Generally, separating the critical section from the rest of the logic in that function is not worth the pain because the taking the lock for the entire duration of the function is not expensive.

This refactor adds a new type, `SentryProfilingState`, that centralizes all logic for reading/writing the shared data structures. This has the additional benefit that you no longer need a `processBacktrace` function that takes many parameters where the lifetime of those parameters is unclear (this will be helpful for fixing the memory leaks in a subsequent PR).

All notable changes have been commented on inline.

## :green_heart: How did you test it?

Ran all existing tests, they pass.

## :pencil: Checklist

You have to check all boxes before merging:

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Sources/Sentry/SentryProfiler.mm [hunk 1]
```diff
@@ -49,7 +50,6 @@
 #    endif
 
 const int kSentryProfilerFrequencyHz = 101;
-NSString *const kTestStringConst = @"test";
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:52]: This constant was unused so I removed it

File: Sources/Sentry/SentryProfiler.mm [hunk 2]
```diff
@@ -421,12 +293,150 @@
...
+- (void)mutate:(void (^)(SentryProfilingMutableState *))block
+{
+    NSParameterAssert(block);
+    std::lock_guard<std::mutex> l(_lock);
+    block(_mutableState);
+}
+
+- (void)appendBacktrace:(const Backtrace &)backtrace {
+    [self mutate:^(SentryProfilingMutableState *state) {
+        const auto threadID = sentry_stringForUInt64(backtrace.threadMetadata.threadID);
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:337]: The logic inside `mutate` is essentially unchanged from the original logic of `processBacktrace` prior to the changes in #3018. The original code is here: https://github.com/getsentry/sentry-cocoa/blob/b7bd71993d0f05fd733eb36c665b7048a476e4fd/Sources/Sentry/SentryProfiler.mm#L91-L160

We now take a single lock to make the entire mutation which simplifies the code, there is no fine grained locking.

File: Sources/Sentry/SentryProfiler.mm [hunk 3]
```diff
@@ -421,12 +293,150 @@
...
+    
+    return @{
+        @"profile": @{
+            @"samples": samples,
+            @"stacks": stacks,
+            @"frames": frames,
+            @"thread_metadata": threadMetadata,
+            @"queue_metadata": queueMetadata
+        }
+    };
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm]: This logic is preserved from the changes in #3018

File: Sources/Sentry/SentryProfiler.mm [hunk 4]
```diff
@@ -421,12 +293,150 @@
...
+
+@end
+
 @implementation SentryProfiler {
-    NSMutableDictionary<NSString *, id> *_profileData;
+    SentryProfilingState *_state;
     std::shared_ptr<SamplingProfiler> _profiler;
     SentryMetricProfiler *_metricProfiler;
     SentryDebugImageProvider *_debugImageProvider;
-    thread::TIDType _mainThreadID;
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:429]: This was unused

File: Tests/SentryProfilerTests/SentryProfilerTests.mm [hunk 5]
```diff
@@ -60,15 +60,7 @@ - (void)testProfilerCanBeInitializedOffMainThread
...
-    const auto resolvedThreadMetadata =
-        [NSMutableDictionary<NSString *, NSMutableDictionary *> dictionary];
-    const auto resolvedQueueMetadata = [NSMutableDictionary<NSString *, NSDictionary *> dictionary];
-    const auto resolvedStacks = [NSMutableArray<NSArray<NSNumber *> *> array];
-    const auto resolvedSamples = [NSMutableArray<SentrySample *> array];
-    const auto resolvedFrames = [NSMutableArray<NSDictionary<NSString *, id> *> array];
-    const auto frameIndexLookup = [NSMutableDictionary<NSString *, NSNumber *> dictionary];
-    const auto stackIndexLookup = [NSMutableDictionary<NSString *, NSNumber *> dictionary];
-
+    SentryProfilingState *state = [[SentryProfilingState alloc] init];
```
Comment (indragiek) [inline Tests/SentryProfilerTests/SentryProfilerTests.mm:63]: `SentryProfilingState` simplifies these tests

File: Sources/Sentry/SentryProfiler.mm [hunk 6]
```diff
@@ -79,108 +79,6 @@
...
-            const auto frame = [NSMutableDictionary<NSString *, id> dictionary];
-            frame[@"instruction_addr"] = instructionAddress;
-#    if defined(DEBUG)
-            const auto symbols
-                = backtrace_symbols(reinterpret_cast<void *const *>(backtrace.addresses.data()),
-                    static_cast<int>(backtrace.addresses.size()));
-            frame[@"function"] = parseBacktraceSymbolsFunctionName(symbols[backtraceAddressIdx]);
-#    endif
-            [stack addObject:@(frames.count)];
-            frameIndexLookup[instructionAddress] = @(frames.count);
```
Comment (indragiek) [inline Sources/Sentry/SentryProfiler.mm:111]: In the old implementation the logic was changed to use the `framesToAdd` array as a temporary container before applying the append all at once in the critical section, but that means that `frames.count` will now be incorrect here.

**Review by brustolin — APPROVED** [review]
LGTM

File: Tests/SentryProfilerTests/SentryProfilerTests.mm [hunk 7]
```diff
@@ -153,9 +145,8 @@ - (void)testProfilerMutationDuringSlicing
     mutateExpectation.expectedFulfillmentCount = operations;
 
     void (^mutateBlock)(void) = ^(void) {
-        processBacktrace(backtrace, resolvedThreadMetadata, resolvedQueueMetadata, resolvedSamples,
-            resolvedStacks, resolvedFrames, frameIndexLookup, stackIndexLookup);
-        [mutateExpectation fulfill];
+        [state mutate:^(
+            __unused SentryProfilingMutableState *mutableState) { [mutateExpectation fulfill]; }];
```
Comment (armcknight) [inline Tests/SentryProfilerTests/SentryProfilerTests.mm:149]: @indragiek I just noticed this change when going to write some new tests. Is the last backtrace created above ([here](https://github.com/getsentry/sentry-cocoa/pull/3035/files#diff-bfe2315bfeb5647454465dc52d0a0ba09dd139140e3292db7054a329a946e22bL149)) no longer appended to the profile data?

**Pull request merged**